session/grpc.go

```diff
@@ -79,21 +80,42 @@ func monitorHealth(ctx context.Context, cc *grpc.ClientConn, cancelConn func())
 	defer cancelConn()
 	defer cc.Close()
 
-	ticker := time.NewTicker(1 * time.Second)
+	ticker := time.NewTicker(5 * time.Second)
 	defer ticker.Stop()
 	healthClient := grpc_health_v1.NewHealthClient(cc)
 
+	hasFailedBefore := false
+	maxHealthcheckDuration := 30 * time.Second
+	lastHealthcheckDuration := time.Duration(0)
+
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			healthcheckStart := time.Now().UTC()
+
+			calculatedTime := maxHealthcheckDuration - time.Duration(float64(lastHealthcheckDuration)*1.5)
```

Comment:
Yes, I didn't mean subtract in my comment. I meant in between `a` and `b`. So effectively `max(maxHealthcheckDuration, time.Duration(float64(lastHealthcheckDuration)*1.5))`

`calculatedTime` -> `timeout`

Reply:
Did you mean `min` instead of `max`? Otherwise `maxHealthcheckDuration` wouldn't actually be an upper bound.

Reply:
`max`. The healthcheck will timeout in 30sec. Except when last healtcheck took (for example) 22 sec, then it would timeout on 22*1.5=33 sec because we already know that the system is behaving slow and give it more time. (I'm not strict on what the actual content values would be).